### PR TITLE
modify

### PR DESCRIPTION
**Pull request by jaywu109**

### CODE REVIEW

File: scribblekitti/config/training.yaml [hunk 1]
```diff
@@ -28,20 +28,21 @@ trainer:
...
   accelerator: 'gpu'
-  sync_batchnorm: False
-  num_sanity_val_steps: 2
-  use_distributed_sampler: True
-  strategy: ddp
+  # check_val_every_n_epoch: 1
+  # sync_batchnorm: False
+  # num_sanity_val_steps: 2
+  # use_distributed_sampler: True
+  # strategy: ddp
```
Comment (jaywu109) [inline scribblekitti/config/training.yaml:39]: strategy & sample 依照实际状况在 `train.py` 中判断 gpu_nums 决定是否使用

File: scribblekitti/train.py [hunk 2]
```diff
@@ -17,17 +14,31 @@
...
+
 class LightningTrainer(pl.LightningModule):
     def __init__(self, config):
         super().__init__()
+
+        seed=123 
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
```
Comment (jaywu109) [inline scribblekitti/train.py:34]: 确保每次模型初始化 weight 相同

File: scribblekitti/train.py [hunk 3]
```diff
@@ -17,17 +14,31 @@
...
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)  
+
         self.config = config
+        self.save_hyperparameters(config)
+        log_folder = config['logger']['root_dir']
+        log_name = config['logger']['name']
+        self.log_folder = f'{log_folder}/{log_name}'
```
Comment (jaywu109) [inline scribblekitti/train.py:40]: 存 `val_miou.txt` 的路径从 config 读取否则每次实验要自己更改可能会忘记

File: scribblekitti/train.py [hunk 4]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
             self.effective_lr = math.sqrt(self.train_batch_size) * self.config["optimizer"]["base_lr"]
 
         print(self.effective_lr)
-        self.save_hyperparameters('config')
+        self.save_hyperparameters({'effective_lr': self.effective_lr})
```
Comment (jaywu109) [inline scribblekitti/train.py:67]: 存取当下使用的 `effecitve_lr`

File: scribblekitti/train.py [hunk 5]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+
+        self.training_step_outputs = {'loss_list': [], 'first_step': None}
```
Comment (jaywu109) [inline scribblekitti/train.py:69]: 避免 training_step & validation_step 中每个 step 都需要 gather 交换信息用

File: scribblekitti/train.py [hunk 6]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+
+    def setup(self, stage=None):
+        """
+        make datasets & seeding each worker separately
+        """
+        ##############################################
+        # NEED TO SET THE SEED SEPARATELY HERE
+        seed = self.global_rank
+       
+        random.seed(seed)
```
Comment (jaywu109) [inline scribblekitti/train.py:80]: 确保不同进程 dataloader 采用不同 seed

File: scribblekitti/train.py [hunk 7]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+        return DataLoader(dataset=self.val_dataset, **self.config['val_dataloader'],
+                          collate_fn=self.collate_fn_BEV, pin_memory=True)
 
     def forward(self, model, fea, pos, bs):
         # fea: list
         # pos: list
         # output_voxel = model([fea.squeeze(0)], [pos.squeeze(0)], bs)
         output_voxel = model(fea, pos, bs)  # [2,20,480,360,32]
-        B = output_voxel.shape[0]     # pos [2,124xxx,3], len(pos)=2
-        outputs = [output_voxel[i, :, pos[i][:,0], pos[i][:,1], pos[i][:,2]] for i in np.arange(B)]
```
Comment (jaywu109) [inline scribblekitti/train.py:59]: 原先的作法会先依照 bs 把 `output_voxel` 转换 point-wise output 后存进 list
接著在下面这里
https://github.com/jaywu109/Scribblekitti_sp2_ddp_modified/blob/106ad1ca21cf808bc55222275c1597ac67c2e47d/scribblekitti/train.py#L73-L75
又会重新把 list 中的 point-wise output 拿出来

如果改成下面的写法，可以减少 ram 存取并减少在 forward 十多使用了一次 for loop
https://github.com/jaywu109/Scribblekitti_sp2_ddp_modified/blob/213ce2aa78562db75b121e036fae353cc961ae44/scribblekitti/train.py#L118-L122

File: scribblekitti/train.py [hunk 8]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
-        return outputs
+        return output_voxel
 
     def training_step(self, batch, batch_idx):
-        if self.global_rank == 0:
```
Comment (jaywu109) [inline scribblekitti/train.py:63]: 这里不确定为啥只在 rank 0 上更新 teacher network，这样其他卡的 teacher net 不更新会让不同卡之间的 weight 不一样吧？

File: scribblekitti/train.py [hunk 9]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
             student_label = student_label_ten[i].unsqueeze(0)
-            student_output, teacher_output = student_output_batch[i].unsqueeze(0), teacher_output_batch[i].unsqueeze(0)
+
+            student_output = student_output_voxel_batch[i, :, student_rpz_ten[i][:,0], student_rpz_ten[i][:,1], student_rpz_ten[i][:,2]].unsqueeze(0)
+            teacher_output = teacher_output_voxel_batch[i, :, teacher_rpz_ten[i][:,0], teacher_rpz_ten[i][:,1], teacher_rpz_ten[i][:,2]].unsqueeze(0)
             loss += self.loss_cl(student_output, teacher_output, student_label) + \
                self.loss_ls(student_output.softmax(1), student_label, ignore=0)
-        train_loss_gather = self.all_gather(loss).mean()
+            
+        self.training_step_outputs['loss_list'].append(loss.cpu().detach())
```
Comment (jaywu109) [inline scribblekitti/train.py:126]: 先把 loss 存下，等到 epoch end 时 gather 再算每个 step 平均后的 loss

File: scribblekitti/train.py [hunk 10]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+        if self.training_step_outputs['first_step'] == None:
+            self.training_step_outputs['first_step'] = self.global_step
 
         if self.global_rank == 0:
-            self.log('train_loss', train_loss_gather, on_epoch=True, prog_bar=True,
-                     batch_size=self.train_batch_size)
+            self.log('train_loss_step', loss, prog_bar=True, rank_zero_only=True, logger=False)
```
Comment (jaywu109) [inline scribblekitti/train.py:131]: 只 log rank 0 step loss 确保没有 nan 即可

File: scribblekitti/train.py [hunk 11]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
         if self.global_rank == 0:
-            self.log('train_loss', train_loss_gather, on_epoch=True, prog_bar=True,
-                     batch_size=self.train_batch_size)
+            self.log('train_loss_step', loss, prog_bar=True, rank_zero_only=True, logger=False) 
 
         return loss
+    
+    def on_train_epoch_end(self):
+        loss_list = torch.stack(self.training_step_outputs['loss_list']).to(self.device)
+        loss_list = self.all_gather(loss_list)
```
Comment (jaywu109) [inline scribblekitti/train.py:137]: epoch end 后只需要 gather 一次

File: scribblekitti/train.py [hunk 12]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+        if self.gpu_num > 1:
+            loss_step_mean = loss_list.mean(dim=0)
+        else:
+            loss_step_mean = loss_list
+
+        if self.global_rank == 0:
+            first_step = self.training_step_outputs['first_step']
+            for loss, step in zip(loss_step_mean, range(first_step, first_step+len(loss_step_mean))):
```
Comment (jaywu109) [inline scribblekitti/train.py:145]: 计算每个 step gather 後的 loss

File: scribblekitti/train.py [hunk 13]
```diff
@@ -48,56 +64,110 @@ def __init__(self, config):
...
+                self.logger.experiment.add_scalar('train_loss_step', loss, step)
```
Comment (jaywu109) [inline scribblekitti/train.py:146]: 计算完后存进 logger

File: scribblekitti/train.py [hunk 14]
```diff
@@ -106,18 +176,23 @@ def validation_step(self, batch, batch_idx):
...
 
     def on_validation_epoch_end(self):
+        loss_mean = torch.stack(self.eval_step_outputs['loss_list']).to(self.device).mean()
+        loss_mean = self.all_gather(loss_mean).mean()
+
         if self.gpu_num > 1:
             student_cm_compute = self.all_gather(self.student_cm.compute().float()).mean(0)
             teacher_cm_compute = self.all_gather(self.teacher_cm.compute().float()).mean(0)
+        else:
+            student_cm_compute = self.student_cm.compute().float()
```
Comment (jaywu109) [inline scribblekitti/train.py:189]: 没有 else 下的这两行代码的会导致在单卡时报错

File: scribblekitti/train.py [hunk 15]
```diff
@@ -126,7 +201,8 @@ def on_validation_epoch_end(self):
                 self.best_iou = np.nan_to_num(teacher_iou) * 100
             self.log('val_best_miou', self.best_miou, on_epoch=True, prog_bar=True, logger=False)
 
-            file = open('/home/jzhang2297/weaksup/scribblekitti/logs/val_iou.txt', "a")
+            file_path = os.path.join(self.log_folder, 'val_iou.txt')
```
Comment (jaywu109) [inline scribblekitti/train.py:204]: 使用 config 给定

File: scribblekitti/train.py [hunk 16]
```diff
@@ -192,20 +259,22 @@ def _load_dataset_info(self) -> None:
...
-        stu_fea2stack = [d['student'][1] for d in data]
-        stu_label2stack = [d['student'][2] for d in data]
+        stu_rpz2stack = []
+        stu_fea2stack = []
+        stu_label2stack = []
+        tea_rpz2stack = []
+        tea_fea2stack = []
+        tea_label2stack = []
+
+        for d in data:
```
Comment (jaywu109) [inline scribblekitti/train.py:270]: 改成只使用一次 for loop，否则原本的作法会跑多次 for loop

File: scribblekitti/train.py [hunk 17]
```diff
@@ -220,11 +289,37 @@ def collate_fn_BEV(self, data):
 
     config = yaml.safe_load(open(args.config_path, 'r'))
     config['dataset'].update(yaml.safe_load(open(args.dataset_config_path, 'r')))
-    wandb_logger = WandbLogger(config=config,
-                               save_dir=config['trainer']['default_root_dir'],
-                               **config['logger'])
+
+    gpus = config['trainer']['devices']
+    gpu_num = len(gpus)
+    print(gpus)
```
Comment (jaywu109) [inline scribblekitti/train.py:295]: strategy & sampler 在此设置

File: scribblekitti/train.py [hunk 18]
```diff
@@ -220,11 +289,37 @@ def collate_fn_BEV(self, data):
...
+    log_folder = config['logger']['root_dir']
+    log_name = config['logger']['name']
+    os.makedirs(f'{log_folder}/{log_name}', exist_ok=True)
+    tb_logger = pl_loggers.TensorBoardLogger(save_dir=log_folder,
+                                             name=log_name,
+                                             default_hp_metric=False)
+    checkpoint = pl.callbacks.ModelCheckpoint(dirpath=f'{log_folder}/{log_name}', filename='{epoch}-{val_teacher_miou:.2f}',
+                                                save_last=True, monitor='val_teacher_miou', mode='max', save_top_k=3)    
+    
+    backup_dir = os.path.join(log_folder, log_name, 'backup_files_%s' % str(datetime.datetime.now().strftime('%Y-%m-%d_%H-%M')))
```
Comment (jaywu109) [inline scribblekitti/train.py:312]: 因为主要文件每次实验或许都会进行些微改动，复制当下使用的代码实验结束后要检查问题较小

**Review by Flora279 — APPROVED** [review]
